### PR TITLE
[FIX] viewport: viewport maxSize should consider the client size

### PR DESCRIPTION
**Pull request by rrahir**
Task 3204720

## Description:

description of this task, what is implemented and why it is implemented that way.

Odoo task ID : [3204720](https://www.odoo.com/web#id=3204720&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_lt("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: tests/plugins/header_visibility.test.ts [hunk 1]
```diff
@@ -95,10 +96,12 @@ describe("Hide Columns", () => {
         },
       ],
     });
+    model.dispatch("RESIZE_SHEETVIEW", { width: DEFAULT_CELL_WIDTH, height: 1000 });
     const sheet = model.getters.getActiveSheet();
     const dimensions = model.getters.getMainViewportRect();
+    let dimensions2: Rect;
     hideColumns(model, ["B", "C", "D"], sheet.id);
-    let dimensions2 = model.getters.getMainViewportRect();
+    dimensions2 = model.getters.getMainViewportRect();
```
Comment (hokolomopo) [inline tests/plugins/header_visibility.test.ts:104]: Why this change?

File: src/helpers/internal_viewport.ts [hunk 2]
```diff
@@ -72,23 +72,29 @@ export class InternalViewport {
...
-        ? Math.max(DEFAULT_CELL_HEIGHT + 5, Math.min(topRowSize, this.height - lastRowSize))
-        : 0);
+    let width = lastColEnd - this.offsetCorrectionX;
+    if (this.canScrollHorizontally) {
+      width += Math.max(
+        DEFAULT_CELL_WIDTH,
+        Math.min(leftColSize, this.width - lastColSize),
+        this.width - width
+      );
+    }
```
Comment (hokolomopo) [inline src/helpers/internal_viewport.ts:88]: I'd add a comment to explain what all those computation mean. W/o prior knowledge of the viewport, it's really hard to get what's going on.

File: src/helpers/internal_viewport.ts [hunk 3]
```diff
@@ -72,23 +72,29 @@ export class InternalViewport {
...
+
+    let height = lastRowEnd - this.offsetCorrectionY;
+    if (this.canScrollVertically) {
+      height += Math.max(
+        DEFAULT_CELL_HEIGHT + 5,
```
Comment (hokolomopo) [inline src/helpers/internal_viewport.ts:93]: +5 ?

File: tests/plugins/grid_manipulation.test.ts [hunk 4]
```diff
@@ -1021,9 +1021,15 @@ describe("Rows", () => {
...
-      expect(dimensions).toMatchObject({ width: 192, height: 124 });
       expect(model.getters.getNumberRows(sheetId)).toBe(6);
+      const dimensions = model.getters.getMainViewportRect();
+      expect(dimensions).toMatchObject({ width: 1000, height: 1000 });
+      model.dispatch("RESIZE_SHEETVIEW", {
+        width: DEFAULT_CELL_WIDTH,
+        height: DEFAULT_CELL_HEIGHT,
+      });
+      const newDimensions = model.getters.getMainViewportRect();
+      expect(newDimensions).toMatchObject({ width: 192, height: 124 });
```
Comment (hokolomopo) [inline tests/plugins/grid_manipulation.test.ts:1032]: I'm gonna trust you on that, but why resizing the sheet to `{DEFAULT_CELL_WIDTH, DEFAULT_CELL_HEIGT}` leads to a size of `{192, 124}` ?

**Review by hokolomopo — COMMENTED** [review]
Viewport computations are really obscure and confusing. We should at least add some comments to explain what's going on IMO

File: src/helpers/internal_viewport.ts [hunk 5]
```diff
@@ -72,23 +66,29 @@ export class InternalViewport {
...
-      this.offsetCorrectionY +
-      (this.canScrollVertically
-        ? Math.max(DEFAULT_CELL_HEIGHT + 5, Math.min(topRowSize, this.height - lastRowSize))
-        : 0);
+    let width = lastColEnd - this.offsetCorrectionX;
+    if (this.canScrollHorizontally) {
+      width += Math.max(
+        DEFAULT_CELL_WIDTH, // leave some minimal space to let the user know they scrolled all the way
+        Math.min(leftColSize, this.clientWidth - lastColSize), // Add pixels that allows the snapping at maximum horizontal scroll
+        this.clientWidth - width // if the viewsport grid size is smaller than its client width, return client width
```
Comment (LucasLefevre) [inline src/helpers/internal_viewport.ts:80]: In this case, it just results in `this.clientWidth`, right ? (+width - width cancel out)
I think I'd just to `Math.max(width, this.clientWidth)` at the end to try simplify this already not so easy max(min).
The new formatting with comment is definitely better.

Comment (rrahir) [reply src/helpers/internal_viewport.ts:80]: agreed, I was a bit too eager to save some lines ^^

File: src/helpers/internal_viewport.ts [hunk 6]
```diff
@@ -23,8 +13,8 @@ export class InternalViewport {
   offsetScrollbarY: Pixel;
   canScrollVertically: boolean;
   canScrollHorizontally: boolean;
-  width: Pixel;
-  height: Pixel;
+  clientWidth: Pixel;
```
[the remark this answers is not in the document]
Comment (rrahir) [reply src/helpers/internal_viewport.ts:16]: When I discussed this with Adrien, it turns out he did not really understand the meaning of that prop. the constructor argument `sizeInGrid` is a bit more explicit but yet fails to really describe what we want to; 
It is the visible size (width/height) of the viewport inside the grid, hence the idea to reuse the term clientWidth,  maybe visibleWidth/visibleHeight would be better ?

**Comment by rrahir** [comment]
@LucasLefevre bump

File: src/helpers/internal_viewport.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (rrahir) [reply src/helpers/internal_viewport.ts]: we are definitely mixing the two concepts. the reason we did this when introducing the frozen pane is that the concept of pane size seemed to only make sense for some viewports and not the others (it's not 100%  but in a real life situation, it is)
And since we only display the content of the viewport, we decided to gowith the vieport naming.

note:: while we named the class viewport, is HAS to be aware of it maximum size because of the scrolling stuff

A more accurate way to describ things would be to stick with the "Pane" as we define them in the core data and have the pane expose a viewport (for instance to use the getVisibleRect getters). but the sutation would be reversed. the Pane would HAVE to be aware of its visible part in the client (DOM) viewport.   Do you think it's more appropriate ? (I kinda do tbh)

**Comment by rrahir** [comment]
@LucasLefevre updated with viewportWidth,Height :)

**Review by LucasLefevre — COMMENTED** [review]
robodoo r+
